Fix flakyness

Despite prior work, still some flakyness. Source is 

dist-packages/rosmaster/master_api.py:208: `xmlrpcapi(api).publisherUpdate('/master', topic, pub_uris)`. pub_uris is an empty list. This publisherUpdate method is an autogenerated method from an xml rpc api somewhere else in ROS, so debugging this is mildly annoying. See this error message:

```
[rosmaster.threadpool][ERROR] 2017-05-14 18:56:01,541: Traceback (most recent call last):
  File "/opt/ros/indigo/lib/python2.7/dist-packages/rosmaster/threadpool.py", line 218, in run
    result = cmd(*args)
  File "/opt/ros/indigo/lib/python2.7/dist-packages/rosmaster/master_api.py", line 208, in publisher_update_task
    xmlrpcapi(api).publisherUpdate('/master', topic, pub_uris)
  File "/usr/lib/python2.7/xmlrpclib.py", line 1233, in __call__
    return self.__send(self.__name, args)
  File "/usr/lib/python2.7/xmlrpclib.py", line 1587, in __request
    verbose=self.__verbose
  File "/usr/lib/python2.7/xmlrpclib.py", line 1273, in request
    return self.single_request(host, handler, request_body, verbose)
  File "/usr/lib/python2.7/xmlrpclib.py", line 1301, in single_request
    self.send_content(h, request_body)
  File "/usr/lib/python2.7/xmlrpclib.py", line 1448, in send_content
    connection.endheaders(request_body)
  File "/usr/lib/python2.7/httplib.py", line 975, in endheaders
    self._send_output(message_body)
  File "/usr/lib/python2.7/httplib.py", line 835, in _send_output
    self.send(msg)
  File "/usr/lib/python2.7/httplib.py", line 797, in send
    self.connect()
  File "/usr/lib/python2.7/httplib.py", line 778, in connect
    self.timeout, self.source_address)
  File "/usr/lib/python2.7/socket.py", line 571, in create_connection
    raise err
```

I believe what's happening is that add_one is publishing an updated value and the nodes that subscribe to it don't exist/have unregistered themselves for some reason.

--- a/pyrostest/launch_tools.py
+++ b/pyrostest/launch_tools.py
@@ -5,25 +5,15 @@ import os
 import sys
 import time
 
-import rosgraph
 import roslaunch
-import rosnode
 
+import pyrostest.rostest_utils
 
 class RosLaunchException(Exception):
     """An Exception thrown when the with_launch_file decorator is misused.
     """
     pass
 
-
-def my_get_node_names(namespace=None, uri='http://localhost:11311'):
-    """Monkeypatches get_node_names with a non-default ROS_MASTER_URI.
-    """
-    old_master = rosgraph.Master
-    rosgraph.Master = functools.partial(rosgraph.Master, master_uri=uri)
-    nodenames = rosnode.get_node_names(namespace=namespace)
-    rosgraph.Master = old_master
-    return nodenames
 
 
 class ROSLauncher(roslaunch.scriptapi.ROSLaunch):
@@ -162,7 +152,8 @@ def launch_node(package, name, namespace=None):
             # Beware this is a bit of a hack, and will currently not work if we
             # want to run more than 1 node with the same name.
             while not any(nn.split('/')[-1].startswith(name.replace('.', '_'))
-                    for nn in my_get_node_names(uri=self.rosmaster_uri)):
+                    for nn in
+                    pyrostest.rostest_utils.my_get_node_names(uri=self.rosmaster_uri)):
                 time.sleep(.1)
             try:
                 temp = func(self)

--- a/pyrostest/ros_test.py
+++ b/pyrostest/ros_test.py
@@ -51,7 +51,7 @@ class MockPublisher(object):
         """
         value.serialize(self.proc.stdin)
         # Just long enough to prevent out of order
-        time.sleep(.2)
+        time.sleep(.5)
 
     def kill(self):
         """Kills the publisher ros node process.
@@ -63,14 +63,14 @@ class MockSubscriber(object):
     """Wrapper around a node used for testing.
     """
 
-    def __init__(self, topic, msg_type, timeout=10):
+    def __init__(self, topic, msg_type, timeout):
         self.timeout = timeout
         self.topic = topic
         self.msg_type = msg_type
         self.killed = False
 
         this_dir = os.path.dirname(os.path.abspath(__file__))
-        location = os.path.join(this_dir, 'listener.py')
+        location = os.path.join(this_dir, 'subscriber.py')
         self.proc = subprocess.Popen([location,
             pickle.dumps((topic, msg_type))], stdout=subprocess.PIPE)
         self._message = None
@@ -129,13 +129,13 @@ class RosTest(unittest.TestCase):
     __metaclass__ = pyrostest.rostest_utils.RosTestMeta
 
     @contextlib.contextmanager
-    def check_topic(self, topic, rosmsg_type, timeout=10):
+    def check_topic(self, topic, rosmsg_type, timeout=15):
         """Context manager that monitors a rostopic and gets a message sent to
         it.
         """
         test_node = MockSubscriber(topic, rosmsg_type, timeout=timeout)
         no_ns = topic.split('/')[-1]
-        while not any(nn.split('/')[-1].startswith( ''.join(['mock_listen_',
+        while not any(nn.split('/')[-1].startswith( ''.join(['mock_subscribe_',
             no_ns])) for nn in rosnode.get_node_names()):
             time.sleep(.1)
 

--- a/pyrostest/ros_test.py
+++ b/pyrostest/ros_test.py
@@ -127,7 +127,7 @@ class RosTest(unittest.TestCase):
     __metaclass__ = pyrostest.rostest_utils.RosTestMeta
 
     @contextlib.contextmanager
-    def check_topic(self, topic, rosmsg_type, timeout=15):
+    def check_topic(self, topic, rosmsg_type, timeout=10):
         """Context manager that monitors a rostopic and gets a message sent to
         it.
         """

--- a/pyrostest/ros_test.py
+++ b/pyrostest/ros_test.py
@@ -15,8 +15,6 @@ import threading
 from StringIO import StringIO
 import unittest
 
-import rosnode
-
 import pyrostest.rostest_utils
 
 
@@ -49,9 +47,8 @@ class MockPublisher(object):
     def send(self, value):
         """Sends data to be publoshed to the mocked topic.
         """
-        value.serialize(self.proc.stdin)
-        # Just long enough to prevent out of order
         time.sleep(.5)
+        value.serialize(self.proc.stdin)
 
     def kill(self):
         """Kills the publisher ros node process.
@@ -87,7 +84,7 @@ class MockSubscriber(object):
 
         Makes sure you've actually recieved a message before providing one.
         """
-        if  not self._message:
+        if not self._message:
             msg = self.msg_type()
             sio = StringIO()
             msg.serialize(sio)
@@ -112,8 +109,9 @@ class MockSubscriber(object):
                 timer.cancel()
 
             if self.killed:
-                raise TimeoutError('No message published on {} within {} '
-                        'seconds'.format(self.topic, self.timeout))
+                raise TimeoutError('No message published to {} within {} '
+                        'seconds. Needed {} bytes, got {}.'.format(self.topic,
+                            self.timeout, sio.len, len(data)))
             msg.deserialize(data)
             self._message = msg
         return self._message
@@ -136,7 +134,8 @@ class RosTest(unittest.TestCase):
         test_node = MockSubscriber(topic, rosmsg_type, timeout=timeout)
         no_ns = topic.split('/')[-1]
         while not any(nn.split('/')[-1].startswith( ''.join(['mock_subscribe_',
-            no_ns])) for nn in rosnode.get_node_names()):
+            no_ns])) for nn in
+            pyrostest.rostest_utils.my_get_node_names(uri=self.rosmaster_uri)):
             time.sleep(.1)
 
         try:
@@ -152,7 +151,8 @@ class RosTest(unittest.TestCase):
         pub = MockPublisher(topic, rosmsg_type, queue_size)
         no_ns = topic.split('/')[-1]
         while not any(nn.split('/')[-1].startswith( ''.join(['mock_publish_',
-            no_ns])) for nn in rosnode.get_node_names()):
+            no_ns])) for nn in
+            pyrostest.rostest_utils.my_get_node_names(uri=self.rosmaster_uri)):
             time.sleep(.1)
         try:
             yield pub

--- a/pyrostest/rostest_utils.py
+++ b/pyrostest/rostest_utils.py
@@ -2,13 +2,25 @@
 """
 
 import os
+import functools
 import random
 import socket
 import subprocess
 import time
-import unittest
 
 import psutil
+import rosgraph
+import rosnode
+
+
+def my_get_node_names(namespace=None, uri='http://localhost:11311'):
+    """Monkeypatches get_node_names with a non-default ROS_MASTER_URI.
+    """
+    old_master = rosgraph.Master
+    rosgraph.Master = functools.partial(rosgraph.Master, master_uri=uri)
+    nodenames = rosnode.get_node_names(namespace=namespace)
+    rosgraph.Master = old_master
+    return nodenames
 
 
 def rand_port():

--- a/pyrostest/subscriber.py
+++ b/pyrostest/subscriber.py
@@ -24,6 +24,7 @@ def subscribe(topic, msg_type):
         """A default callback for our subscriber.
         """
         message.serialize(sys.stdout)
+        rospy.signal_shutdown('message sent')
 
     rospy.Subscriber(topic, msg_type, listener_cb)
     rospy.spin()

--- a/pyrostest/subscriber.py
+++ b/pyrostest/subscriber.py
@@ -14,16 +14,17 @@ import sys
 import rospy
 
 
-def listen(topic, msg_type):
+
+def subscribe(topic, msg_type):
     """A function to isolate some of the subscriber logic.
     """
-    rospy.init_node('mock_listen_'+topic.split('/')[-1], anonymous=True,
+    rospy.init_node('mock_subscribe_'+topic.split('/')[-1], anonymous=True,
             disable_signals=True)
     def listener_cb(message):
         """A default callback for our subscriber.
         """
         message.serialize(sys.stdout)
-        rospy.signal_shutdown('message sent')
+
     rospy.Subscriber(topic, msg_type, listener_cb)
     rospy.spin()
 
@@ -31,4 +32,4 @@ def listen(topic, msg_type):
 
 if __name__ == '__main__':
     ROSTOPIC, ROSMSG_TYPE = pickle.loads(sys.argv[1])
-    listen(ROSTOPIC, ROSMSG_TYPE)
+    subscribe(ROSTOPIC, ROSMSG_TYPE)

--- a/setup.py
+++ b/setup.py
@@ -21,7 +21,7 @@ test = [
 
 setup(
     name='pyrostest',
-    version='0.0.5',
+    version='0.1.0',
     description='The most lit ros testing framework',
     long_description=long_description,
     packages=['pyrostest'],

--- a/setup.py
+++ b/setup.py
@@ -21,7 +21,7 @@ test = [
 
 setup(
     name='pyrostest',
-    version='0.0.4',
+    version='0.0.5',
     description='The most lit ros testing framework',
     long_description=long_description,
     packages=['pyrostest'],

--- a/test/pyrostest/add_one.py
+++ b/test/pyrostest/add_one.py
@@ -10,7 +10,7 @@ def add_one_cb(val):
 
 def add_one():
     rospy.init_node('adder', anonymous=True)
-    rospy.Subscriber('/subber', Int32, add_one_cb)
+    rospy.Subscriber('/pub_val', Int32, add_one_cb)
     rospy.spin()
 
 if __name__ == '__main__':

--- a/test/pyrostest/test_add_one.py
+++ b/test/pyrostest/test_add_one.py
@@ -5,7 +5,7 @@ import pyrostest
 class TestAddOne(pyrostest.RosTest):
     @pyrostest.launch_node('pyrostest', 'add_one.py')
     def test_add_one_node(self):
-        with self.mock_pub('/subber', Int32, queue_size=0) as sub:
+        with self.mock_pub('/pub_val', Int32, queue_size=0) as sub:
             with self.check_topic('/pyrostest/add_one', Int32) as out:
                 sub.send(Int32(5))
                 assert out.message.data == 6

